text/0000-freeze-uninit.md

```diff
@@ -0,0 +1,217 @@
+
+Any initialized byte value chosen nondeterminstically may be chosen arbitrarily by the implementation (and, in particular, the implementation is permitted to deliberately pick a value that will violate the validity invariant of `T` or otherwise cause undefined behaviour, if such a choice is available). 
+
+
+## Library Functions
+
+### `core::ptr::read_freeze` 
+```rust
+// in module core::ptr
+pub unsafe fn read_freeze<T>(ptr: *const T) -> T{
+    core::intrinsics::read_freeze(ptr)
+}
+```
+
+The `core::ptr::read_freeze` function directly invokes the `read_freeze` intrinsic and returns the result. 
+
+The majority of the operation is detailed [above](#language-primitive), so the definition is not repeated here.
+
+The `read_freeze` functions on `*const T` and `*mut T` are method versions of the free function and may be defined in the same manner.
+
+### `MaybeUninit::<T>::freeze`
+
+```rust 
+// in module core::mem
+impl<T> MaybeUninit<T>{
+    pub fn freeze(self) -> Self{
+        unsafe{core::ptr::read_freeze(core::ptr::addr_of!(self))}
+    }
+}
+```
+
+`MaybeUninit::freeze` is a safe version of the `read_freeze` function, defined in terms of the language intrinsic. It is safe because the value is kept as `MaybeUninit<T>`, which is trivially valid. 
+To be used, Rust code would need to unsafely call `MaybeUninit::assume_init` and assert that the frozen bytes are valid for `T`.
+
+The remaining behaviour of the function is equivalent to the `read_freeze` intrinsic.
+
+### Example Alternative Implementation
+
+If an implementation decides to define `MaybeUninit::freeze` (or equivalent) as the compiler intrinsic, it is possible to implement `core::ptr::read_freeze` as follows:
+```rust 
+pub unsafe fn read_freeze(ptr: *const T) -> T{
+    struct AlignedBytes([u8; core::mem::size_of::<T>()], [T;0]);
+    
+    let val = ptr.cast::<MaybeUninit<AlignedBytes>>().read();
+    
+    core::mem::transmute(val.freeze())
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The main drawbacks that have been identified so far:
+* It is potentially [considered desireable](https://rust-lang.zulipchat.com/#narrow/stream/136281-t-opsem/topic/Arguments.20for.20freeze/near/377333420) to maintain the property that sound (e.g. correct) code cannot meaningfully read uninitialized memory
+    * It is generally noted that safe/unsafe is not a security or privilege boundary, and it's fully possible to write unsound code (either deliberately or inadvertanly) that performs the read. If the use of uninitialized memory is within the threat model of a library that, for example, handles cryptographic secrets, that library should take additional steps to santize memory that contains those secrets.
+    * Undefined behaviour does not prevent malicious code from accessing any memory it physically can.
+
+
+# Rationale and alternatives
```

Comment:
An alternative that is being considered by `zerocopy` (cc @joshlf) is something like:
```rust
/// Initializes possibly-uninitialized bytes in `v` to `0`.
///
/// # Safety
///
/// Unsafe code may depend on uninitialized bytes in `v` being
/// initialized to `0`. Note, however, that the initialization
/// of padding bytes is not preserved on move.
// TODO: Restrict `v` to non-dyn pointers.
fn initialize<T: ?Sized>(v: &mut T) {}
```
This alternative mitigates some of the listed Drawbacks of this RFC's proposal: Rust would retain the property that sound code does not read uninitialized memory.

Reply:
To implement `initialize` as-written (i.e. specifically initializing padding bytes to *zero*, not just some arbitrary value, and not modifying non-padding bytes), this would require compiler support to know *where* padding bytes are in an arbitrary type, and as the comment mentions it definitely wouldn't work for `dyn Trait` in general.

If would also that require compiler support to guarantee "better-than-O(size_of_val)" performance, even for arbitrary initialization instead of zero-initialization.

Reply:
This doesn't work with `dyn Trait` (with the `?Sized` bound there - though that is mentioned), nor on types that allow uninit bytes that aren't in padding bytes. This does not function for the `MaybeUinit<Scalar>` case.

Reply:
@zachs18:

> To implement `initialize` as-written (i.e. specifically initializing padding bytes to _zero_, not just some arbitrary value, and not modifying non-padding bytes), this would require compiler support to know _where_ padding bytes are in an arbitrary type, and as the comment mentions it definitely wouldn't work for `dyn Trait` in general.

The compiler *does* know where padding bytes are for arbitrary types. That's the entire shtick of Project Safe Transmute, and it's also part of how miri is able to validate transmutations at runtime. 

> If would also that require compiler support to guarantee "better-than-O(size_of_val)" performance, even for arbitrary initialization instead of zero-initialization.

I don't follow why that would be the case. Both `initialize` and `freeze` are O(size_of_val). 

---

@chorman0773:

> This doesn't work with `dyn Trait` (with the `?Sized` bound there - though that is mentioned)

As I understand it, neither `initialize` nor `freeze` would work on such types. At any rate, I haven't seen any demand from folks doing zero-copy parsing to support `dyn Trait`. (If anyone reading this has a use-case for that, please let us know!)

> nor on types that allow uninit bytes that aren't in padding bytes.

As documented, it would overwrite them. It would probably be useful to provide `initialize_padding` that *only* overwrote padding bytes, too.

> This does not function for the `MaybeUinit<Scalar>` case.

Can you say more about this?

Reply:
@chorman0773, I think @RalfJung has identified a compelling reason why we'd want `freeze` over just `initialize` (though for many use cases of PST, an `initialize` is perhaps easier to reason about)! I'd still like to hear about the `MaybeUinit<Scalar>` case, but I think there's probably enough here in this comment thread to robustly flesh out `initialize` in the *Alternatives* section and defend why `freeze` is still desirable.

Reply:
@jswrenn 
> > If would also that require compiler support to guarantee "better-than-O(size_of_val)" performance, even for arbitrary initialization instead of zero-initialization.
> 
> I don't follow why that would be the case. Both `initialize` and `freeze` are O(size_of_val).

Earlier, you said

> The performance cost of this is linear with respect to the number of padding bytes in a type.

If a type `T` has no padding bytes it follows that `[T]` has no padding bytes and thus `initialize::<[T]>`'s performance does not (asymptotically) depend on the length of the slice, but `size_of_val` does depend on the length of the slice (for non-ZST `T`).

Reply:
Perhaps I should have used a different term than "compiler support". Obviously the RFC's proposed freeze operation would also require compiler support. My point was that this operation "does more" than the RFC's proposed operation and thus requires the compiler to use more information about the type being operated on to implement it, whereas the freeze operation semantically only needs one "magic operation" `MaybeUninit<u8> -> u8` to specify.

Reply:
@zachs18 Let me rephrase: Is there a performance bound here that you would think would be difficult to guarantee? For `T` with no uninit bytes, the `initialize` is a no-op. For `T: ?Sized` with uninit bytes, the number of writes is linear with respect to `size_of_val(&t)`. In the worst case (all uninit bytes), the number of writes would almost certainly be equal to `size_of_val(&t)`.

Contrast this with by-value `freeze`, which (baring unguaranteed compiler optimizations) always requires moving `size_of_val(&t)` bytes.

Reply:
> ```rust
> /// Unsafe code may depend on uninitialized bytes in `v` being
> /// initialized to `0`.
> ```

I think that's unimplementable because sometimes LLVM doesn't know if bytes are uninitialized (e.g. calling an external function), so it can't know which bytes to zero. `freeze` without zeroing works because when LLVM doesn't know if a byte is uninitialized, it just uses the current machine-level value of that byte, which is always implementable.

Reply:
> As documented, it would overwrite them. It would probably be useful to provide initialize_padding that only overwrote padding bytes, too.

In order for this to work, either the compiler would have to know and preserve which bytes are uninit (violating the refinement that an uninitialized byte to any initialized value which is not implementable), or it would have to set *every* byte to 0.

Reply:
@programmerjake The semantics of `initialize` and `freeze` are different. The compiler does, in fact, know whether bytes in a type are possibly uninit, because the compiler knows where it has placed padding and `union` fields. However, the compiler cannot know, statically, whether union-introduced uninit bytes have been overwritten at runtime, and so `initialize` would either need to unconditionally overwrite union-introduced uninit bytes (`initialize`), or leave them untouched (`initialize_padding`).

`freeze` has the advantage that it preserves the runtime values union-introduced uninit bytes. However, for the sort of zero-copy (de)serialization done by consumers of bytemuck and zerocopy, this property isn't necessarily important (it's unusual that you'd cast initialized bytes into a `MaybeUninit`).

I don't claim that `initialize` covers all use-cases of `freeze`, but I do claim that it satisfies many of the use-cases relevant to Project Safe Transmute, and does so without some of the drawbacks of `freeze`:
- it retains the language property that runtime uninitialized bytes cannot be read in sound code
- it works on slice DSTs

Reply:
`initialize` isn't really an alternative IMO so much as it is just a different operation. As far as I can tell, `initalize` is for when you have `&mut T` (with already valid `T`), want `&mut [u8]` preserving the validity of `T`, are okay with branching code for initializing through `enum`, and are okay overwriting any and all `union` bytes.

Without an additional bound that `union` isn't used, I don't see this being particularly useful, even for safe transmute, although PST is also likely the most likely to be able to enforce that bound. With the bound it's useful for getting to `&[u8]` which can then be cast further, but in the abstract I can't see any usage that wants this operation not exclusively as a way to bypass using difficult to express "uninit at the same (or more) byte offsets" bounds.

There's no need for `initialize` to accept `?Sized`, either. Just have `initialize` for sized types and a separate `initialize_slice` for slices. I suppose that doesn't work for custom slice tail DST, but that's also status quo for most API.

Reply:
Yeah, I'll agree that `initialize` is a completely different operation. It's probably worth at least tangentially mentioning in the Alternatives section, though. Initialize doesn't do a whole lot of good for use cases 2 and 3, though.